Fixed #35581 -- Updated django.core.mail to Python's modern email API.

medmunds opened the pull request:
#### Trac ticket number

ticket-35581

#### Branch description

Switch django.core.mail from using Python's "legacy email API" to its newer "modern email API."

~~I believe this PR is ready for review, but I've marked it "draft" for now. It should not be merged until fixes land for two security issues in Python's modern email API: python/cpython#80222 and python/cpython#121284. (I've included tests in this PR that will fail until they've been addressed.)~~ [Edit: The security issues [have been addressed](#issuecomment-2663947020).]

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- n/a I have attached screenshots in both light and dark modes for any UI changes.

#### Notes

This PR includes a few changes not covered in the original django-developers discussion. More details [in the ticket](https://code.djangoproject.com/ticket/35581#comment:19), but briefly:

* Deprecated passing legacy MIMEBase objects to `EmailMessage.attach()`; added support for modern MIMEPart objects instead.
* Dropped support for the undocumented `EmailMessage.mixed_subtype` and `EmailMultiAlternatives.alternative_subtype` properties and for using legacy Charset objects in the undocumented `EmailMessage.encoding` property. (As these are undocumented, dropped without deprecation. But attempting to use any of them will result in an error.)

I've split the changes into separate commits that may be easier to review individually. The first three commits update tests for django.core.mail only, without changing the implementation. These updated tests should pass whether django.core.mail uses Python's legacy or modern email API to generate messages.

The <del>four</del> <ins>five</ins> commits are (more details in the commit messages):

1. Adds trailing newlines to all text content (bodies and attachments) in mail tests. This is because Python's modern API forces trailing newlines in text (python/cpython#121515).

   In the real world, most email text content probably already has a newline at the end, or it won't cause a problem to add one. Unfortunately, a lot of our test cases include variations on `body="Content"` and `assertEqual(generated_body, "Content")` without trailing newlines, which break with the modern APIs. I've updated those tests to use trailing newlines for all text.

2. Prepares the mail tests for switching to the modern API by reducing dependencies on specific implementation details. (E.g., checking that a message sent with non-ASCII content will parse correctly to the original content at the receiving end, rather than checking for specific encodings and byte sequences.)

   This commit also adds comments to describe some tests that will be relocated or removed in the modern email update. (And then commit 4 removes those comments.)

3. Adds tests for the two Python issues python/cpython#80222 and python/cpython#121284.

   These pass with the legacy API, but currently fail with the modern one. <del>I don't believe there's any practical workaround Django could implement, so until they're fixed in Python the next commit must not be merged.</del>

4. Finally switches to the modern email API. (You could review this commit in isolation if you just want to see what is actually changing.)

5. [Edit: added.] Implements a workaround for one of the security issues. (And the other is fixed in Python 3.12.9 and 3.13.2.)

medmunds committed:
Fixed #35581 -- Updated django.core.mail to Python's modern email API.

- Changed EmailMessage.message() to construct a "modern email API"
  email.message.EmailMessage and added policy keyword arg.
- Added support for modern MIMEPart objects in EmailMessage.attach()
  (and EmailMessage constructor, EmailMessage.attachments list).
- Updated SMTP EmailBackend to use modern email.policy.SMTP.

Deprecated:
- Attaching MIMEBase objects (replace with MIMEPart)
- BadHeaderError (modern email uses ValueError)
- SafeMIMEText, SafeMIMEMultipart (unnecessary for modern email)
- django.core.mail.forbid_multi_line_headers()
  (undocumented, but exposed via `__all__` and in wide use)
- django.core.mail.message.sanitize_address()
  (undocumented, but in wide use)

Removed without deprecation (all undocumented):
- EmailMessage.mixed_subtype
- EmailMultiAlternatives.alternative_subtype
- Support for setting (undocumented) EmailMessage.encoding property
  to a legacy email.charset.Charset object

Related changes:
- Dropped tests for incorrect RFC 2047 encoding of non-ASCII email
  address localparts. This is specifically prohibited by RFC 2047, and
  not supported by any known MTA or email client. (Python still
  mis-applies encoded-word to non-ASCII localparts, but it is a bug that
  may be fixed in the future.)
- Added tests that try to discourage using Python's legacy email APIs
  in future updates to django.core.mail.

knyghty commented:
@medmunds I see the first issue you linked is now fixed, just not backported everywhere yet. For https://github.com/python/cpython/issues/121284, I notice that on the issue it says "modern or legacy" - if that's the case, should it block this work? It seems like it is _already_ an issue.

medmunds commented:
> I see the first issue you linked is now fixed, just not backported everywhere yet.

Yes. And the auto backports are failing—I'll try to take a look at those. But since the 3.12 backport was merged, and the next version of Django requires Python 3.12+, this is no longer a blocker.

> For [python/cpython#121284](https://github.com/python/cpython/issues/121284), I notice that on the issue it says "modern or legacy" - if that's the case, should it block this work? It seems like it is _already_ an issue.

The policy names get confusing: if you serialize a legacy Message using its default legacy `policy.compat32`, that does *not* have the bug. And that's what Django currently does, so it's not vulnerable. But if you serialize any legacy Message or modern EmailMessage using the *modern* `policy.default` (like this PR), that will be vulnerable to address header injection.

medmunds commented:
OK, I believe the security issues are resolved and this PR is ready for review.

* The fix for python/cpython#80222 has been released in Python 3.12.9 and 3.13.2.
* I was able to find a reasonable workaround for python/cpython#121284, implemented as a fifth commit in this PR.

medmunds commented:
(The Windows test is running Python 3.13.2, and succeeds. The Jenkins tests seem to be using Python 3.13.0, and will keep failing until upgraded to a current version of Python 3.12 or 3.13.)

medmunds commented:
Cpython has now also merged a fix for fix for python/cpython#121284. It should show up in Python 3.12.10 and 3.13.3. So both Python security issues will be fixed in all (latest) supported Python releases before this is likely to get merged.

I'm not sure whether to keep or remove the workaround.

knyghty commented:
@medmunds I suspect it's safe to remove, as this won't land until 6.0, which only supports 3.12+, and we state that we only support the latest patch release.

medmunds commented:
@knyghty thanks. I'll remove the workaround once those Pythons are released and django-ci's runners get updated. I think it makes sense to leave the tests in place, just in case.

(Incidentally, both security fixes have been backported to Python 3.9, 3.10, and 3.11 as well.)

sarahboyce commented:
I agree it's safe to remove the workaround. It may even be worth dropping the test to confirm the python behavior is fixed (as we try not to test Python behavior but I understand why it was added in this case)

sarahboyce commented on tests/mail/tests.py:
```diff
@@ -960,9 +962,11 @@ def test_attach_mime_image(self):
-    def test_attach_mime_image_in_constructor(self):
-        image = MIMEImage(b"\x89PNG...", "png")
-        image["Content-Disposition"] = "attachment; filename=test.png"
+    def test_attach_mime_part_in_constructor(self):
```
Note that I tried to follow the deprecation removal instructions and the following tests failed:
- `mail.tests.MailTests.test_attach_mime_part_in_constructor`
- `mail.tests.MailTests.test_decoded_attachment_text_MIMEPart`

Do these need comments on how to be adjusted? Or is something wrong with the other instructions do you think?

medmunds replied:
This code in EmailMessage.\_\_init\_\_() was prone to confusing interpretation. I've split the compound isinstance() to clarify:

```patch
         self.attachments = []
         if attachments:
             for attachment in attachments:
-                # RemovedInDjango70Warning: MIMEBase attachment support.
-                if isinstance(attachment, (MIMEBase, email.message.MIMEPart)):
+                if isinstance(attachment, email.message.MIMEPart):
+                    self.attach(attachment)
+                elif isinstance(attachment, MIMEBase):
+                    # RemovedInDjango70Warning: MIMEBase attachment support.
                     self.attach(attachment)
                 else:
                     self.attach(*attachment)
```

medmunds replied:
I've opened draft PR #19577 on top of this to verify the deprecation removals.

sarahboyce commented on django/core/mail/message.py:
```diff
@@ -362,68 +217,69 @@ def attach_file(self, path, mimetype=None):
-            # attachments must not be base64 encoded.
+    def _add_attachment(self, msg, filename, content, mimetype):
+        encoding = self.encoding or settings.DEFAULT_CHARSET
+        if mimetype is None:
+            mimetype = DEFAULT_ATTACHMENT_MIME_TYPE
+        maintype, subtype = mimetype.split("/", 1)
+        # See email.contentmanager.set_content() docs for the cases here.
+        if maintype == "text":
+            # For text/*, content must be str, and maintype cannot be provided.
+            if isinstance(content, bytes):
+                content = content.decode()
+            msg.add_attachment(
+                content, subtype=subtype, filename=filename, charset=encoding
+            )
```
```suggestion
        maintype, subtype = mimetype.split("/", 1)
        # See email.contentmanager.set_content() docs for the cases here.
        if maintype == "text":
            msg.add_attachment(
                content, subtype=subtype, filename=filename, charset=encoding
            )
```
I can update this to the above without test failures

medmunds replied:
So I've gone back and forth on this a couple of times. There's definitely a missing test case and a slight bug in my implementation.

I've concluded that there's also an existing bug where setting EmailMessage.attachments doesn't work the same as EmailMessage.attach() for utf-8 encoded text/* content.

I've prepended a commit that fixes that (and adds test cases). With that, the diff in this part of the code is clearer, and your suggested update would break the new tests.

medmunds replied:
Done: ticket-36478.

If that's not accepted, the alternative would be to check for this case here and raise a useful error. (Passing bytes with a text/* mimetype to EmailMessage.add_attachment causes a cryptic error.)

sarahboyce reviewed (commented):
I have made https://github.com/django/django/pull/19573 which resolves conflicts and drops the Python workaround in case this is useful
I had a quick look at the test coverage and deprecation instructions also :+1:

medmunds commented on django/core/mail/backends/smtp.py:
```diff
@@ -145,18 +148,52 @@ def _send(self, email_message):
+            parsed = AddressHeader.value_parser(address)
+        except (AssertionError, HeaderParseError, IndexError, TypeError, ValueError):
+            # (Don't include the parser error message. It's generally not helpful.)
+            raise ValueError(f"Invalid address {address!r}") from None
```
I believe that entire except clause may be left over from Python ~3.9–3.11 support. Or possibly from an earlier version of this PR that used a private Python email API here—code that was [borrowed from the current sanitize_address()](https://github.com/django/django/blob/bc1bfe12b613334bd625aeb36fd44af96d186c10/django/core/mail/message.py#L87). (That I went to the trouble to recompose the error message suggests I did actually encounter at least some of these exceptions at some point.) 

In any case, I can't find any need for protecting AddressHeader.value_parser() under Python 3.12–3.13 in the current code. Removed.

medmunds commented on django/core/mail/message.py:
```diff
@@ -486,24 +368,28 @@ def attach_alternative(self, content, mimetype):
+            for alternative in self.alternatives:
+                maintype, subtype = alternative.mimetype.split("/", 1)
+                content = alternative.content
+                if maintype == "text":
+                    if isinstance(content, bytes):
+                        content = content.decode()
+                    msg.add_alternative(content, subtype=subtype, charset=encoding)
+                else:
+                    content = force_bytes(content, encoding=encoding, strings_only=True)
+                    msg.add_alternative(content, maintype=maintype, subtype=subtype)
```
I'm resolving this by adding a test case for add_alternative with non-text mimetypes.

The, er, alternative would be to allow only text/* alternatives and add code here (and a test) that raises an error if non-string content or non-text/* mimetypes are passed to Django's attach_alternative().

Pre-PR, Django does not support non-text alternatives, and raises a cryptic error on them somewhere in SafeMIME.

Python's email package *does* support any type of alternative (as do the email specs), but raises an equally cryptic error about unexpected arguments from deep within its contentmanager if you try to call add_alternative() on bytes content with a subtype and charset (or on string content with a maintype).

sarahboyce reviewed (commented):
Thank you for the updates @medmunds :star: I really appreciate your persist progress here

sarahboyce commented:
@apollo13 :wave: I noticed you were involved in the [Django developer discussion](https://groups.google.com/g/django-developers/c/2zf9GQtjdIk) of this change and done some work Django email work in the past. If you have any insights or opinions on this PR, they're very welcome :+1:

medmunds commented on tests/mail/tests.py:
```diff
@@ -692,15 +723,19 @@ def test_encoding(self):
+
+        # Check that body is actually encoded with iso-8859-1.
+        msg_bytes = message.as_bytes()
+        if message["Content-Transfer-Encoding"] == "quoted-printable":
+            self.assertIn(b"Firstname S=FCrname is a great guy.", msg_bytes)
+        elif message["Content-Transfer-Encoding"] == "8bit":
+            self.assertIn(b"Firstname S\xfc", msg_bytes)
+        else:
+            self.fail("Unexpected Content-Transfer-Encoding")
```
I think this is carryover from implementation-neutral tests that worked with either legacy or modern APIs. It can be simplified now.

medmunds commented on tests/mail/tests.py:
```diff
@@ -2622,3 +2774,65 @@ def test_fail_silently_on_connection_error(self):
         self.backend.open()
+
+
+# Check whether python/cpython#128110 has been fixed by seeing if space
+# between encoded-words is ignored (as required by RFC 2047 section 6.2).
```
> It's strange to me to only see references to working around a Python bug within our test suite (as we use `message_from_bytes` in our code)

Ah, I'd been thinking message_from_bytes was only used in the test suite. But you're correct, there is a *single* use in django.core.mail. You can add a message/rfc822 attachment with EmailMessage.attach("filename.eml", b"From: ...\r\nTo: ...\r\n...", "message/rfc822"). We use message_from_bytes to convert the content to a message attachment.

The Python bug is not a security issue, and I don't believe should be a blocker for this PR.

Unlike that one relatively narrow case in django.core.mail, the tests use message_from_bytes a lot, and go out of their way to test cases impacted by the bug, so a workaround in the tests made sense to me.

There a few options:
* Pull the workaround into django.core.mail. I'm confident the workaround is sufficiently good for purposes of the test cases; I'm less confident about treating it as a supported feature.
* Wait for the cpython fix to land. (Realistically, that could happen next month, next year, or five years from now. Also, since it's not a security issue, backports will be limited.)
* Recognize that Python modern email has bugs too, and that they're *different* bugs from legacy email. (I mentioned that [as a risk](https://groups.google.com/g/django-developers/c/2zf9GQtjdIk/m/GWjimue9FAAJ#:~:text=There%27s%20probably%20some,Email%20is%20complicated.) in the original discussion.) This is one of those bugs. If a similar non-security bug were reported against django.core.mail due to the legacy APIs, I *suspect* we would close it as wontfix, point to the upstream issue, and suggest the user implement their own workarounds. (I could even preemptively open and close a Django ticket for this issue once this PR is merged 😄 )

medmunds replied:
(Pulled workaround into a separate commit, but kept it test-only.)

sarahboyce commented:
The fix for ticket-36478 is merged
Thank you for the replies to the previous comments, they all sound reasonable to me. I have a holiday coming up but will get back to this when I'm back (mid-July ish)

apollo13 commented:
> @apollo13 👋 I noticed you were involved in the [Django developer discussion](https://groups.google.com/g/django-developers/c/2zf9GQtjdIk) of this change and done some work Django email work in the past. If you have any insights or opinions on this PR, they're very welcome 👍

I fear I do not have the time to do so currently. That said, the thoughtfulness and  planning Mike has been putting into this (from his earlier changes and design specs) is so stellar that I do not think there is any chance that we'd end up with anything worse than we have now (I actually believe the end result will be massively better).

If bugs occur afterwards feel free to blame me for not reviewing ;)

medmunds commented:
@sarahboyce I pushed fixup commits for a missing deprecation-cleanup comment and some long comments/docstrings.

sarahboyce commented on django/core/mail/message.py:
```diff
@@ -199,7 +31,10 @@ class EmailMessage:
 
     content_subtype = "plain"
-    mixed_subtype = "mixed"
+
+    # Undocumented charset to use for text/* message bodies
+    # and attachments. Defaults to settings.DEFAULT_CHARSET.
+    # (Should rename to `charset` if it will be documented.)
```
```suggestion
```
I'm not sure this comment helps me, are we refering to encoding here?

sarahboyce replied:
I don't think we have a convention

medmunds commented on docs/topics/email.txt:
```diff
@@ -362,12 +379,18 @@ The class has the following methods:
-  override this method to put the content you want into the MIME object.
+* ``message(policy=email.policy.default)`` constructs and returns a Python
+  :class:`email.message.EmailMessage` object using the specified policy,
+  representing the message to be sent.
```
I expanded that to try to provide some additional context:

```rst
  The keyword argument ``policy`` allows specifying the set of rules for
  updating and serializing the representation of the message. It must be an
  :py:mod:`email.policy.Policy <email.policy>` object. While its default value
  :py:data:`email.policy.default` is generally useful, in certain cases you may
  want to use :py:data:`~email.policy.SMTP`, :py:data:`~email.policy.SMTPUTF8`
  or a custom policy. (For example, Django's built-in SMTP EmailBackend uses
  the SMTP policy to ensure ``\r\n`` line endings as required by the SMTP
  protocol.)
```

(Python's module-level docs for email.policy provide a better introductory explanation of the email.policy.Policy class than the class API docs.)

medmunds commented on django/core/mail/message.py:
```diff
@@ -1,196 +1,28 @@
 from pathlib import Path
 
 from django.conf import settings
+from django.core.mail._deprecated import (  # NOQA: F401; RemovedInDjango70Warning
+    forbid_multi_line_headers,
+    sanitize_address,
+)
```
Correct, forbid_multi_line_headers() and sanitize_address() are both *undocumented* but commonly used. This is here to avoid directly breaking code that uses them. They now issue deprecation warnings on use. sanitize_address() was only available in `django.core.mail.message`; forbid_multi_line_headers() was exposed both there and in `django.core.mail`.[^1]

All of `BadHeaderError`, `SafeMIMEText`, and `SafeMIMEMultipart` are *documented* as being in `django.core.mail`, not `django.core.mail.message`. They now get deprecation warnings *on import* from `django.core.mail.__init__`. (And don't provide warnings on use. That wouldn't even be possible for `BadHeaderError`. I think I looked at adding a warning "on use" in the SafeMIME\* constructors and either decided it wasn't helpful or ran into some problem with it.)

I think I'm OK with breaking imports of (documented) APIs from *undocumented* paths. But if we think it's important to preserve this, we'd need to duplicate the deprecate-on-import logic from module root in message.

[^1]: In addition to the deprecation warning on use, forbid_multi_line_headers() will also issue a deprecation warning on *import* from `django.core.mail`, but not from `django.core.mail.message`, because the code was easier that way. Since it's undocumented anyway, that inconsistency didn't really bother me. I was aiming for *at least one* deprecation warning.

medmunds commented on django/core/mail/message.py:
```diff
@@ -277,18 +116,19 @@ def message(self):
             # Avoid headers handled above.
             if name.lower() not in {"from", "to", "cc", "reply-to"}:
-                msg[name] = value
+                msg[name] = force_str(value, strings_only=True)
```
Hmm, I would have sworn the original code was `msg[name] = force_str(value)` and I just added the `strings_only=True` part (and test_datetime_in_date_header()). But the diff doesn't lie.

This is an unreported bug in the existing code. I'll open a new issue. Here's the failing test case:
```python
    def test_lazy_headers(self):
        message = EmailMessage(
            headers={"List-Unsubscribe": gettext_lazy("mailto:unsub@example.com")},
        ).message()
        self.assertEqual(
            message.get_all("List-Unsubscribe"), ["mailto:unsub@example.com"]
        )
```

medmunds replied:
Actually, it works in existing code, but it's a missing test. Prepended a commit.

medmunds committed:
Refs #35581 -- Added missing test for lazy email headers.

EmailMessage is intended to support lazy strings in any header field
(via coercion to `str` in forbid_multi_line_headers() called from
SafeMIMEMessage/Text/Multipart.__setitem__).

sarahboyce commented:
Note that I have pushed up changes to reduce the diff (discussed with Natalia and have a slight preference to avoid moving functions/classes from files to make it easier to review what has just "moved" and what has "changed and moved")

I am planning to do a final comb through today/tomorrow and hope to land this very soon :+1:

sarahboyce reviewed (approved):
:partying_face: :woman_dancing: :confetti_ball: 

This looks good to me! Thank you for the consistent effort on this ticket :clap: lovely work

I will mark as "ReadyForCheckin" and if we don't get further feedback, land in a couple of days

medmunds commented:
🎉 🎉 🎉 

#19661 will create a non-trivial conflict (because of the code that moved from message.py to \_deprecated.py). I'll rebase and resolve after that's merged.
